Add support for SRV record for ClusterIP services

aswinsuryan opened the pull request:
Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

aswinsuryan committed:
Add support for SRV record for ClusterIP services

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

tpantelis commented on pkg/service/controller.go:
```diff
@@ -84,24 +87,41 @@ func (c *Controller) Stop() {
+	}
 	if svc.Spec.Type == v1.ServiceTypeClusterIP && svc.Spec.ClusterIP != "" {
-		return svc.Spec.ClusterIP, true
+		return record, true
```
You can move this check above L105:

```
if svc.Spec.Type != v1.ServiceTypeClusterIP || svc.Spec.ClusterIP == "" {
    return nil, false
}
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/serviceimport/map.go:
```diff
@@ -23,24 +23,28 @@ import (
+type DNSRecord struct {
+	IP   string
+	Port []mcsv1a1.ServicePort
+}
```
New line.

```suggestion
}

```

aswinsuryan replied:
Done.

tpantelis commented on pkg/serviceimport/map.go:
```diff
@@ -61,16 +65,16 @@ func (m *Map) selectIP(queue []clusterInfo, counter *uint64, name, namespace str
-	checkEndpoint func(string, string, string) bool) (ip string, found, isLocal bool) {
-	clusterIPs, queue, counter, isHeadless := func() (map[string]string, []clusterInfo, *uint64, bool) {
+	checkEndpoint func(string, string, string) bool) (record *DNSRecord, found, isLocal bool) {
+	clusterIPs, queue, counter, isHeadless := func() (map[string]*DNSRecord, []clusterInfo, *uint64, bool) {
```
```suggestion
	dnsRecords, queue, counter, isHeadless := func() (map[string]*DNSRecord, []clusterInfo, *uint64, bool) {
```

aswinsuryan replied:
Done.

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+}
+
+func (lh *Lighthouse) createSRVRecords(record *serviceimport.DNSRecord, state request.Request, pReq recordRequest, zone string) []dns.RR {
+	reqPorts := make([]mcsv1a1.ServicePort, 0)
```
No need to initialize if not needed.

```suggestion
	var reqPorts []mcsv1a1.ServicePort
```

aswinsuryan replied:
Done.

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+
+func (lh *Lighthouse) createSRVRecords(record *serviceimport.DNSRecord, state request.Request, pReq recordRequest, zone string) []dns.RR {
+	reqPorts := make([]mcsv1a1.ServicePort, 0)
+	records := make([]dns.RR, 0)
```
Move this right before L110:

```suggestion
	records := make([]dns.RR, len(reqPorts))
```

aswinsuryan replied:
Done.

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+}
+
+func (lh *Lighthouse) createARecords(ips []string, state request.Request) []dns.RR {
+	records := make([]dns.RR, 0)
```
```suggestion
	records := make([]dns.RR, 0, len(ips))
```

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -91,12 +152,18 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
 
 	records := make([]dns.RR, 0)
```
```suggestion
	var records []dns.RR
```

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -91,12 +152,18 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+		records = lh.createARecords(ips, state)
+	} else if state.QType() == dns.TypeSRV {
+		records = lh.createSRVRecords(record, state, pReq, zone)
+		if records == nil {
```
Perhaps this should be moved after L165?

```suggestion
		if len(records) {
```

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+func (lh *Lighthouse) createSRVRecords(record *serviceimport.DNSRecord, state request.Request, pReq recordRequest, zone string) []dns.RR {
+	reqPorts := make([]mcsv1a1.ServicePort, 0)
+	records := make([]dns.RR, 0)
+	portReqKey := pReq.port + pReq.protocol
+	if portReqKey == "" {
+		reqPorts = record.Port
+	}
+
+	for _, port := range record.Port {
+		portKey := strings.ToLower(port.Name) + strings.ToLower(string(port.Protocol))
+		log.Debugf("The current port Name %q SRV require port Name %q", portKey, portReqKey)
+
+		if portKey == portReqKey {
+			reqPorts = append(reqPorts, port)
+		}
+	}
```
If no port is requested then we're unnecessarily looping on `record.Port`.  Also what if` pReq.port` is empty but `pReq.protocol` isn't?  How about: 

```suggestion
	if pReq.port == "" {
		reqPorts = record.Port
	} else {
	    log.Debugf("Requested port %q, protocol %q for SRV", pReq.port, pReq.protocol)
	    for _, port := range record.Port {
		name := strings.ToLower(port.Name)
		protocol := strings.ToLower(string(port.Protocol))
		
		log.Debugf("Checking port %q, protocol %q", name, protocol)

		if name == pReq.port && protocol ==  pReq.protocol {
			reqPorts = append(reqPorts, port)
		}
	    }
	}
```

aswinsuryan replied:
Done.

tpantelis commented on plugin/lighthouse/handler.go:
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
+		target = pReq.cluster + "." + target
+	}
+
+	for _, ports := range reqPorts {
```
```suggestion
	for _, port := range reqPorts {
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/serviceimport/map.go:
```diff
@@ -23,24 +23,28 @@ import (
 
+type DNSRecord struct {
+	IP   string
+	Port []mcsv1a1.ServicePort
```
```suggestion
	Ports []mcsv1a1.ServicePort
```

aswinsuryan replied:
Done.

tpantelis commented on plugin/lighthouse/parse.go:
```diff
@@ -115,3 +107,38 @@ func (r recordRequest) String() string {
+		case 1: // cluster and hostname
+			r.cluster = segs[count]
+			r.hostname = segs[count-1]
+			r.hostname = segs[count-1]
```
Duplicate line.

```suggestion
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/service/controller.go:
```diff
@@ -84,24 +87,41 @@ func (c *Controller) Stop() {
 
 	svc := obj.(*v1.Service)
 
+	var mcsServicePort []mcsv1a1.ServicePort
```
```suggestion
	var mcsServicePorts []mcsv1a1.ServicePort
```

aswinsuryan replied:
Done.

aswinsuryan committed:
Apply review comments

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

vthapar commented:
Can we split SRV handler into a separate file? Would be even better to use a plugin/interface type design similar to way etc/kubernetes plugins are written, but that can be done later.

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -88,16 +88,16 @@ func (m *MockClusterStatus) LocalClusterID() string {
 
 type MockLocalServices struct {
-	LocalServicesMap map[string]string
+	LocalServicesMap map[string]*serviceimport.DNSRecord
```
Also add UTs for SRV queries.

aswinsuryan replied:
Added UTs

mangelajo reviewed (dismissed):
test

mangelajo commented:
> test

sorry, please ignore :-/

aswinsuryan committed:
Apply review comments

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

aswinsuryan commented:
> Can we split SRV handler into a separate file? Would be even better to use a plugin/interface type design similar to way etc/kubernetes plugins are written, but that can be done later.

Moved record creation to a different file.

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -172,46 +204,87 @@ func testWithoutFallback() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace2 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
```
```suggestion
					test.SRV(service1 + "." + namespace2 + ".svc.clusterset.local.    5    IN    SRV 0 50 "+ strconv.itoa(portNumber1) +
```

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -362,12 +477,24 @@ func testClusterStatus() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
```suggestion
					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 " + strconv.itoa(portNumber2)
```

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -379,6 +506,17 @@ func testClusterStatus() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
```
Same as above. Use `strconv.itoa(portNumber1)`

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -608,6 +799,17 @@ func testLocalService() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Same as above

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
same as above

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Same as above

vthapar commented on plugin/lighthouse/handler_test.go:
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
+					test.SRV(clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
+						clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local."),
+					test.SRV(clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
+						clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local."),
```
Same as above

aswinsuryan committed:
Apply review comments

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

aswinsuryan committed:
Merge branch 'devel' into srv-record

aswinsuryan commented:
test
